### ISSUE TITLE
Support returning a function when using the PhpDriver

### ISSUE THREAD

**Issue by arjanfrans** [doctrine/persistence#200]
When defining ORM mapping using the `php` driver there is `$metadata` variable floating around in the mapping file. IDE's complain about type hinting. 

It would be nice to be able to return a static function that gets the $metadata variable as an argument, so then you can write the mapping in php like this:

```
// config/orm/mapping/App.Entity.User.php

<?php

use Doctrine\ORM\Mapping\Builder\ClassMetadataBuilder;
use Doctrine\ORM\Mapping\ClassMetadata;

return static function (ClassMetadata $metadata): void
{
    $builder = new ClassMetadataBuilder($metadata);
    
    // ...
};
```

I modified the PHP driver to be able to do this (`lib/Doctrine/Persistence/Mapping/Driver/PHPDriver.php`):

```
// ...
    protected function loadMappingFile($file)
    {
        $metadata = $this->metadata;
        $mapping = include $file;
    
        if (is_callable($mapping)) {
            $mapping($metadata);
        }

        return [$metadata->getName() => $metadata];
    }
// ...
```

Is there any argument against doing this? The big advantage here is that you at least have a type hint.

**Comment by GromNaN** [comment]
This looks like the [`StaticPHPDriver`](https://github.com/doctrine/persistence/blob/68ea6fbe6080a7dff0e21d832324b18e97c264f9/src/Persistence/Mapping/Driver/StaticPHPDriver.php#L55), that calls automatically the method `loadMetadata` on the mapped class. Is it still relevant now that we have PHP attributes?

**Comment by arjanfrans** [comment]
Yes, still relevant. I do not want loadMetadata on the entity class itself, neither do I want attribute on my entity class, to separate concerns (Domain / DB-Mapping).

**Comment by stof** [comment]
Implemented in https://github.com/doctrine/persistence/pull/450. It will be part of the upcoming 4.2.0 release.
